subiquity/server/controllers/install.py

```diff
@@ -421,22 +424,81 @@ async def run_curtin_step(name, stages, step_config, source=None):
+            if casper_uuid is not None and rp_partuuid is not None:
+                rp = self.app.base_model.partition_by_partuuid(rp_partuuid)
+                if rp is not None:
+                    await self.configure_rp_boot(
+                        context=context, rp=rp, casper_uuid=casper_uuid
+                    )
+
+    async def adjust_rp(self, rp: Partition, mp: Mountpoint) -> str:
+        if self.app.opts.dry_run:
+            return
+        # Once the installer has been copied to the RP, we need to make two
+        # adjustments:
+        #
+        # 1. set a new "casper uuid" so that booting from the install
+        #    media again, or booting from the RP but with the install
+        #    media still attached, does not get confused about which
+        #    device to use as /cdrom.
+        #
+        # 2. add "rp-partuuid" to the kernel command line in grub.cfg
+        #    so that subiquity can identify when it is running from
+        #    the recovery partition and add a reference to it to
+        #    grub.cfg on the target system in that case.
+        grub_cfg_path = mp.p("boot/grub/grub.cfg")
+        new_cfg = []
+        new_casper_uuid = str(uuid.uuid4())
+        cp = await self.app.command_runner.run(
+            ["lsblk", "-n", "-o", "PARTUUID", rp.path], capture=True
+        )
+        rp_uuid = cp.stdout.decode("ascii").strip()
+        with open(grub_cfg_path) as fp:
+            for line in fp:
+                words = shlex.split(line)
+                if words and words[0] == "linux" and "---" in words:
+                    index = words.index("---")
+                    words[index - 1 : index - 1] = [
+                        "uuid=" + new_casper_uuid,
+                        "rp-partuuid=" + rp_uuid,
+                    ]
+                    new_cfg.append(shlex.join(words) + "\n")
+                else:
+                    new_cfg.append(line)
+        with open(grub_cfg_path, "w") as fp:
+            fp.write("".join(new_cfg))
+        for casper_uuid_file in glob.glob(mp.p(".disk/casper-uuid-*")):
+            with open(casper_uuid_file, "w") as fp:
+                fp.write(new_casper_uuid + "\n")
+        return new_casper_uuid
+
+    @with_context(description="configuring grub menu entry for factory reset")
+    async def configure_rp_boot_grub(self, context, rp: Partition, casper_uuid: str):
+        # Add a grub menu entry to boot from the RP
+        cp = await self.app.command_runner.run(
+            ["lsblk", "-n", "-o", "UUID,PARTUUID", rp.path], capture=True
+        )
+        fs_uuid, rp_uuid = cp.stdout.decode("ascii").strip().split()
+        conf = grub_reset_conf.format(
+            HEADER=generate_timestamped_header(),
+            PARTITION=rp.number,
+            FS_UUID=fs_uuid,
+            CASPER_UUID=casper_uuid,
+            RP_UUID=rp_uuid,
+        )
+        with open(self.tpath("etc/grub.d/99_reset"), "w") as fp:
```

Comment:
The following `update-grub` will need to write to the exact file.

I got a failure saying:
```
/usr/sbin/grub-mkconfig: 301: /etc/grub.d/99_reset: Text file busy
```